### PR TITLE
[CPU] Enable fused_attention pattern matcher

### PR DESCRIPTION
**Pull request by Valentine233**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105131
* __->__ #107128
* #104863
* #104693
* #103826
* #104584
* #104583

Feature RFC: https://github.com/pytorch/rfcs/pull/56.

Enable the SDPA graph rewriting for Inductor CPU.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -4766,24 +4784,46 @@ def meta__scaled_dot_product_flash_backward(
...
+            (0, 2, 1, 3),
+            dtype=key.dtype,
+            device=key.device,
+        )
+        grad_v = torch.empty_permuted(
+            (batch_size, num_heads, seq_kv, head_dim),
+            (0, 2, 1, 3),
+            dtype=value.dtype,
+            device=value.device,
+        )
```
Comment (jgong5) [inline torch/_meta_registrations.py:4807]: I think we can further simplify the code if the differences are only the seq length.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_meta_registrations.py (hunk 1 above)
Comment (Valentine233) [reply torch/_meta_registrations.py]: Sure! Modified.

**Review by jgong5 — APPROVED** [review]

**Comment by drisspg** [comment]
cc @eellison on the inductor numbers

**Review by jansel — APPROVED** [review]

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -4694,50 +4694,72 @@ def meta__scaled_dot_product_flash(
...
+            logsumexp,
+            torch.empty((), dtype=torch.int32, device="meta"),
+            torch.empty((), dtype=torch.int32, device="meta"),
+            0,
+            0,
+            torch.empty((), dtype=torch.long, device="meta"),
+            torch.empty((), dtype=torch.long, device="meta"),
+            torch.empty((), dtype=query.dtype, device=query.device),
         )
     else:
```
Comment (eellison) [inline torch/_meta_registrations.py:4718]: nit you dont need the `else` here and the extra indentation

**Review by eellison — APPROVED** [review]

File: torch/_meta_registrations.py (hunk 2 above)
Comment (Valentine233) [reply torch/_meta_registrations.py]: Thanks and modified.

**Pull request closed without merging**